File: typescript/packages/core/src/executors/command-executor.ts
import { spawn } from 'child_process';
import { ToolDefinition } from '../core/schema';
import { MatimoError, ErrorCode } from '../errors/matimo-error';

/**
 * CommandExecutor - Executes shell commands
 * Handles parameter templating, timeouts, and error capture
 */

export class CommandExecutor {
  private cwd?: string;

  constructor(cwd?: string) {
    this.cwd = cwd;
  }

  /**
   * Execute a tool that runs a shell command.
   *
   * @param tool - Tool definition
   * @param params - Tool parameters
   * @param credentials - Optional per-call credential overrides. Keys must match the env-var
   *   names used by the tool (e.g. `SLACK_BOT_TOKEN`). When provided they are merged on top of
   *   `process.env` inside the child process so the spawned script sees them as normal env vars.
   *   Values are never logged. Falls back to the current environment when not provided.
   */
  async execute(
    tool: ToolDefinition,
    params: Record<string, unknown>,
    credentials?: Record<string, string>
  ): Promise<unknown> {
    if (tool.execution.type !== 'command') {
      throw new MatimoError('Tool execution type is not command', ErrorCode.EXECUTION_FAILED, {
        expectedType: 'command',
        actualType: tool.execution.type,
      });
    }

    const { command, args = [], timeout = 30000 } = tool.execution;
    const startTime = Date.now();

    // SECURITY: command must be a fixed executable — never a templated value.
    // Only 'args' may contain {placeholder} tokens.
    if (/\{[^}]+\}/u.test(command)) {
      throw new MatimoError(
        `execution.command must not contain parameter placeholders — only 'args' may be templated. ` +
          `Found: '${command}'. Move the dynamic part into 'args'.`,
        ErrorCode.EXECUTION_FAILED,
        { toolName: tool.name }
      );
    }
    const templatedCommand = command; // Never template the executable
    const templatedArgs = args.map((arg) => this.templateString(arg, params));

    return new Promise((resolve) => {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const spawnOptions: any = {
        stdio: ['pipe', 'pipe', 'pipe'],
        // Merge per-call credentials on top of the current environment so that
        // the spawned process sees them as ordinary env vars. This is safe:
        // values are held only in memory for the duration of the spawn setup
        // and are never written to disk or logged.
        env: credentials ? { ...process.env, ...credentials } : process.env,
      };

      // Set working directory if provided
      if (this.cwd) {
        spawnOptions.cwd = this.cwd;
      }

      const child = spawn(templatedCommand, templatedArgs, spawnOptions);

      let stdout = '';
      let stderr = '';
      let timedOut = false;

      // Set up timeout
      const timer = setTimeout(() => {
        timedOut = true;
        child.kill('SIGTERM');
      }, timeout);

      child.stdout?.on('data', (data) => {
        stdout += data.toString();
      });

      child.stderr?.on('data', (data) => {
        stderr += data.toString();
      });

      child.on('close', (code) => {
        clearTimeout(timer);
        const duration = Date.now() - startTime;

        if (timedOut) {
          resolve({
            success: false,
            error: 'timeout',
            exitCode: -1,
            duration,
          });
        } else {
          const exitCode = code || 0;
          const success = exitCode === 0;

          resolve({
            success,
            stdout: stdout.trim(),
            stderr: stderr.trim(),
            exitCode,
            duration,
          });
        }
      });

      child.on('error', (error) => {
        clearTimeout(timer);
        const duration = Date.now() - startTime;

        resolve({
          success: false,
          error: error.message,
          exitCode: -1,
          duration,
        });
      });
    });
  }

  /**
   * Replace parameter placeholders in a string
   */
  private templateString(str: string, params: Record<string, unknown>): string {
    let result = str;
    for (const [key, value] of Object.entries(params)) {
      const placeholder = `{${key}}`;
      result = result.replace(new RegExp(placeholder, 'g'), String(value));
    }
    return result;
  }
}

export default CommandExecutor;

File: python/packages/core/src/matimo/executors/command_executor.py
"""
Command executor — spawns subprocess for 'type: command' tools.
Mirrors: packages/core/src/executors/command-executor.ts
"""
from __future__ import annotations

import asyncio
import logging
import os
import re
import time
from pathlib import Path
from typing import Any

from matimo.core.models import CommandExecution, ToolDefinition
from matimo.errors import ErrorCode, MatimoError

logger = logging.getLogger("matimo")

_PLACEHOLDER_RE = re.compile(r"\{([^}]+)\}")

_DEFAULT_TIMEOUT_MS = 30_000


class CommandExecutor:
    """
    Executes command tools by spawning subprocesses.
    Mirrors: CommandExecutor in command-executor.ts

    Security:
    - Credentials merged onto env for the child process only — never logged.
    - Shell=False by default to prevent shell injection.
    """

    def __init__(self, cwd: str | None = None) -> None:
        self._cwd = cwd

    async def execute(
        self,
        tool: ToolDefinition,
        params: dict[str, Any],
        credentials: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        """
        Execute a command tool.

        Returns dict: {success, stdout, stderr, exit_code, duration}
        """
        exec_cfg = tool.execution
        if not isinstance(exec_cfg, CommandExecution):
            raise MatimoError(
                f"Tool '{tool.name}' is not a command tool",
                ErrorCode.EXECUTION_FAILED,
                {"tool_name": tool.name, "execution_type": exec_cfg.type},
            )

        # SECURITY: command must be a fixed executable — never a templated value.
        # Only 'args' may contain {placeholder} tokens.
        if _PLACEHOLDER_RE.search(exec_cfg.command):
            raise MatimoError(
                f"execution.command must not contain parameter placeholders — only 'args' may be "
                f"templated. Found: '{exec_cfg.command}'. Move the dynamic part into 'args'.",
                ErrorCode.EXECUTION_FAILED,
                {"tool_name": tool.name},
            )
        command = exec_cfg.command  # Never template the executable
        args = [self._template(a, params) for a in (exec_cfg.args or [])]

        # Build child environment: process env + tool env + credentials (last wins)
        child_env = dict(os.environ)
        if exec_cfg.env:
            child_env.update(exec_cfg.env)
        if credentials:
            child_env.update(credentials)  # SECURITY: not logged

        # Determine working directory
        cwd: str | None = None
        if exec_cfg.cwd:
            cwd = exec_cfg.cwd
        elif self._cwd:
            cwd = self._cwd
        else:
            cwd = str(Path.cwd())

        timeout_s = (exec_cfg.timeout or _DEFAULT_TIMEOUT_MS) / 1000.0
        start = time.monotonic()

        try:
            proc = await asyncio.create_subprocess_exec(
                command,
                *args,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
                env=child_env,
                cwd=cwd,
            )

            try:
                stdout_bytes, stderr_bytes = await asyncio.wait_for(
                    proc.communicate(), timeout=timeout_s
                )
            except TimeoutError:
                proc.kill()
                await proc.communicate()
                duration = time.monotonic() - start
                raise MatimoError(
                    f"Tool '{tool.name}' timed out after {duration:.1f}s",
                    ErrorCode.TIMEOUT,
                    {"tool_name": tool.name, "timeout_ms": exec_cfg.timeout or _DEFAULT_TIMEOUT_MS},
                ) from None

        except MatimoError:
            raise
        except Exception as exc:
            raise MatimoError(
                f"Failed to spawn process for tool '{tool.name}': {exc}",
                ErrorCode.EXECUTION_FAILED,
                {"tool_name": tool.name, "command": command},
                cause=exc,
            ) from exc

        duration = time.monotonic() - start
        exit_code = proc.returncode or 0
        stdout = stdout_bytes.decode("utf-8", errors="replace").strip()
        stderr = stderr_bytes.decode("utf-8", errors="replace").strip()

        return {
            "success": exit_code == 0,
            "stdout": stdout,
            "stderr": stderr,
            "exit_code": exit_code,
            "duration": duration,
        }

    # ------------------------------------------------------------------
    # Templating
    # ------------------------------------------------------------------

    def _template(self, value: str, params: dict[str, Any]) -> str:
        """Replace {placeholder} tokens with param values."""
        def replacer(m: re.Match[str]) -> str:
            name = m.group(1)
            if name in params:
                return str(params[name])
            env_val = os.environ.get(name)
            if env_val is not None:
                return env_val
            raise MatimoError(
                f"Missing value for placeholder '{{{name}}}'",
                ErrorCode.INVALID_PARAMETER,
                {"placeholder": name},
            )

        return _PLACEHOLDER_RE.sub(replacer, value)
